plugins/amqp/src/plugin.cpp

```diff
@@ -279,6 +280,27 @@ class amqp_engine {
     return to_error(status);
   }
 
+  /// Consumes frames from broker, simply for the side effect of processing
+  /// heartbeats implicitly. Required if otherwise no interaction with the
+  /// broker would occur.
+  auto handle_heartbeat() {
+    amqp_frame_t frame;
+    // We impose no timeout, either there is something to read or not. Never
+    // block!
+    struct timeval tv = {0, 0};
+    int status;
+    if (conn_ == nullptr)
+      return;
+    do {
+      status = amqp_simple_wait_frame_noblock(conn_, &frame, &tv);
+      if (AMQP_STATUS_TIMEOUT != status && AMQP_STATUS_OK != status) {
+        TENZIR_WARN("unexpected error while processing heartbeats: {}",
+          amqp_error_string2(status));
+        return;
```

Comment:
If this is a serious error that needs to be propagated upwards in the call stack, you should return a `caf::error` here.

Check out the function `start_consumer()` below on how to spawn & return such an error with `return caf::make_error(....)`, then check out how at lines ~574 the return value `start_consumer` is checked and then turned into a potential diagnostic.

You can do a pretty similar thing in your `weak_run_delayed_loop` lambda function. Try it out and see if that helps you.

Reply:
It is not a serious error. If this happens, we should just note that and try again next time. This is just to cover all bases, and not really something that needs to be handled.

Reply:
Changed to diagnostic output, thanks!